libvast/vast/address_synopsis.hpp

```diff
@@ -48,6 +54,61 @@ class address_synopsis final
   }
 };
 
+/// A synopsis for IP addresses that can be used in situations where
+/// [...]
+// TBD: should this go into a new header `buffered_address_synopsis.hpp`?
+// TBD: should we remove the `final` from address_synopsis and use that as base
+//      class instead?
+template <class HashFunction>
+class buffered_address_synopsis
+  : public bloom_filter_synopsis<address, HashFunction> {
+public:
+  using super = bloom_filter_synopsis<address, HashFunction>;
+
+  /// Constructs an IP address synopsis from an `address_type` and a Bloom
+  /// filter.
+  buffered_address_synopsis(type x, typename super::bloom_filter_type bf)
+    : super{std::move(x), std::move(bf)}, type_(x) {
+  }
+
+  void add(data_view x) override {
+    auto addr_view = caf::get_if<view<address>>(&x);
+    VAST_ASSERT(addr_view);
+    ips_.push_back(*addr_view);
+    super::add(x);
+  }
+
+  synopsis_ptr shrink() const override {
+    // The bloom_filter doesnt store its `p`, so we parse the type to get it.
+    auto old_params = parse_parameters(type_);
+    if (!old_params)
+      return nullptr;
```

Comment:
I think we should just extend our bloom filter implementation for that.

Reply:
I think the design is that the pair (bf, params) describes everything needed?

Reply:
If you only had the bf you wouldn't be able to do the operation here. Adding a copy of the params to the bf would remove the burden of keeping them around separately. The downside is that we would pay for this copy because we can't get rid of `type_`.

Reply:
> The downside is that we would pay for this copy [..]

Not much to pay here; the params are a struct of super-cheap values of size 32bytes. Adding those the Bloom filter wouldn't be an issue. I'd rather decouple this from `type`, which is much "further away" in the coe.

I was just wondering if the outside context keeps the params around anyway. If so, they don't have to go into the Bloom filter. But putting them in there is effectively a no-op in the grand scheme of things.

Reply:
I considered this for a while, but there were still some open questions about whether the synopsis or the bloom filter itself should store this, and how to handle the situations where a bloom filter was constructed raw, without `bloom_filter_params`. So I've left this as-is for now.